File: Problem_1.py
# Time Complexity: O(n)
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

class Solution(object):
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        ans = set()
        seen = set()
        for i in range(len(s) - 9):
            seq = s[i:(i + 10)]
            if seq in seen:
                ans.add(seq)
            seen.add(seq)
        return list(ans)

File: problem1.cpp
/*
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
*/

#include<iostream>
#include<vector>
#include<string>
#include<unordered_map>

using namespace std;

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        const int window_len{10};
        unordered_map<string,int> umap{};
        vector<string> res{};
        int s_len = s.size();
        //cout<<s_len<<endl;
        if(s_len<=window_len) return res;
        for(int i{};i<s_len-window_len+1;++i){
            string temp = s.substr(i,window_len);
            if(umap.find(temp)==umap.end()){
                umap[temp]=0;
            }
            else{
                umap[temp]++;
            }
        }
        for(const auto& pair:umap){
            if(pair.second>0){
                res.push_back(pair.first);
            }
        }
        return res;
    }
};